### DISCUSSION TITLE
Evaluating an optimized function/expression

### DISCUSSION THREAD

**Discussion by aiba** [axkr/symja_android_library#225] [Ideas]
`OptimizeExpression` is awesome!  But I'm having trouble using the Java API to call a function after it's been optimized.

In this example, `f2` becomes the optimized function.  I want to call `f2(1.0)` and was expecting `6.0` as the result, but instead it produces an unevaluated syntax tree.  Could someone kindly nudge me in the right direction to evaluate `f2(1.0)`?  Thank you!

```java
package com.matrixops.scratch;

import org.matheclipse.core.eval.ExprEvaluator;
import org.matheclipse.core.expression.F;
import org.matheclipse.core.interfaces.IAST;
import org.matheclipse.core.interfaces.IExpr;
import org.matheclipse.core.reflection.system.OptimizeExpression;

public class App {
    public static void main(String[] args) throws Exception {
        ExprEvaluator util = new ExprEvaluator(false, (short)100);
        IExpr result;

        // #1+1+(#1+1)*(#1+1)
        IAST f = F.Function(F.Plus(F.Slot1, F.C1, F.Times(F.Plus(F.Slot1, F.C1), F.Plus(F.Slot1, F.C1))));
        result = util.evalFunction(f, "1.0");

        System.out.println("f: "+f.toString());
        System.out.println("f(1.0) = "+result);

        OptimizeExpression opt = new OptimizeExpression();
        IExpr f2 = opt.evaluate(f, util.getEvalEngine());
        result = util.evalFunction(f2, "1.0");  // <--- expected to be 6.0, but instead is a symbolic expression.

        System.out.println("Optimized:");
        System.out.println("f2: "+f2.toString());
        System.out.println("f2(1.0): "+result);
    }
}
```

**Comment by axkr** [comment]
Can you try something like this:

```Java

import org.matheclipse.core.eval.ExprEvaluator;
import org.matheclipse.core.expression.F;
import org.matheclipse.core.expression.S;
import org.matheclipse.core.interfaces.IAST;
import org.matheclipse.core.interfaces.IExpr;

public class App {

  public static void main(String[] args) throws Exception {
    ExprEvaluator util = new ExprEvaluator(false, (short) 100);
    IExpr result;

    // #1+1+(#1+1)*(#1+1)
    IAST f =
        F.Function(F.Plus(F.Slot1, F.C1, F.Times(F.Plus(F.Slot1, F.C1), F.Plus(F.Slot1, F.C1))));
    IExpr function = util.eval(f);
    result = util.evalFunction(function, "1.0");

    System.out.println("f: " + function.toString());
    System.out.println("f(1.0) = " + result);

    IExpr opt = F.unaryAST1(S.OptimizeExpression, function);
    IAST list = (IAST) util.eval(opt);

    System.out.println("Optimized:");
    System.out.println("New expression: " + list.get(1).toString());
    System.out.println("List of variables: " + list.get(2).toString());
  }
}
```

I had to define a function form for `OptimizeExpression` in future version.

**Reply by axkr** [reply]
Added a commit https://github.com/axkr/symja_android_library/commit/d22ac8d4f0e3d84d7a6d620aa60402ae27947a50

Now the example App returns `{v2*v2+v2&,{v1->#1,v2->1+v1}}`  for the `OptimizeExpression` result.

**Reply by aiba** [reply]
This is helpful, thank you!

But my real question is: with the optimized function, how would one use the Java API to evaluate it for a given input value?

I tried: `result = util.evalFunction(list, "1.0");` but that doesn't seem to work.

**Reply by axkr** [reply]
I'm not sure what you want to achieve? 
- Faster evaluation directly in Symja? 
- Rewriting an expression for faster evaluation in another language like C or Java source code?

For performance I could imagine to automatically convert the output in a compiled function:
- https://github.com/axkr/symja_android_library/blob/master/symja_android_library/matheclipse-core/src/main/resources/doc/functions/CompiledFunction.md
- https://github.com/axkr/symja_android_library/blob/master/symja_android_library/matheclipse-core/src/main/resources/doc/functions/Compile.md

Here are some simple possibilities:
```
    IExpr opt = F.unaryAST1(S.OptimizeExpression, function);
    IAST list = (IAST) util.eval(opt);
    IExpr e1 = util.evalFunction(list.get(1), "1.0");
    System.out.println("E1: " + e1.toString());
    IExpr e2 = util.evalFunction(F.Function(list.get(2)), "1.0");
    System.out.println("E2: " + e2.toString());
    result = e1.replaceRepeated((IAST) e2);
    System.out.println("ReplaceRepeated: " + result);
```

**Reply by aiba** [reply]
Yes, thank you for asking more directly what I'm trying to achieve.

Starting from an expression (for example, `(x+1)^2+(x+1)^3`), I want to generate an optimized java function (which contains a dummy variable for the `x+1` subexpression) that takes a double and returns a double.  From java code, if we name this optimized function `f`, I want to be able to call `f(2.0)` from java code and get `36.0` result.

I would appreciate so greatly if you could help out with some example code.  Thank you!

**Reply by axkr** [reply]
I assume you would like to get a String with the pure Java source code, which doesn't depend on Symja?

Something like: 
```
JavaForm((x+1)^2+(x+1)^3, Float) 
```

which splits an expression in the different steps improved by first calling `OptimizeExpression( (x+1)^2+(x+1)^3 ) ` 

- http://matheclipse.org/input?i=JavaForm%28%28x%2B1%29%5E2%2B%28x%2B1%29%5E3%2C%20Float%29
